[/data-oriented-programming]Add ballerina and java samples

pcnfernando opened the pull request:
## Purpose
> $subject

## Goals
> Describe the solutions that this feature/fix will introduce to resolve the problems described above

## Approach
> Describe how you are implementing the solutions. Include an animated GIF or screenshot if the change affects the UI (email documentation@wso2.com to review all UI text). Include a link to a Markdown file or Google doc if the feature write-up is too long to paste here.

## User stories
> Summary of user stories addressed by this change>

## Release note
> Brief description of the new feature or bug fix as it will appear in the release notes

## Documentation
> Link(s) to product documentation that addresses the changes of this PR. If no doc impact, enter “N/A” plus brief explanation of why there’s no doc impact

## Training
> Link to the PR for changes to the training content in https://github.com/wso2/WSO2-Training, if applicable

## Certification
> Type “Sent” when you have provided new/updated certification questions, plus four answers for each question (correct answer highlighted in bold), based on this change. Certification questions/answers should be sent to certification@wso2.com and NOT pasted in this PR. If there is no impact on certification exams, type “N/A” and explain why.

## Marketing
> Link to drafts of marketing content that will describe and promote this feature, including product page changes, technical articles, blog posts, videos, etc., if applicable

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

## Samples
> Provide high-level details about the samples related to this feature

## Related PRs
> List any other related PRs

## Migrations (if applicable)
> Describe migration steps and platforms on which migration has been tested

## Test environment
> List all JDK versions, operating systems, databases, and browser/versions on which this feature/fix was tested
 
## Learning
> Describe the research phase and any blog posts, patterns, libraries, or add-ons you used to solve the problem.

pcnfernando committed:
Integrate optional field access code

pcnfernando commented on working_with_json/main.bal:
```diff
@@ -29,6 +29,13 @@ public function main() returns error?{
+
+    // Results in a nil value if the accessed field is not present.
+    decimal? discountAmount = check itemArr[1]?.discount?.amount;
+    io:println("Discount amount: ", discountAmount.toBalString());
```
Had to use `balString` to print `()`. Any other suggestion @sameerajayasoma ?

pcnfernando committed:
Add working with json java sample

pcnfernando committed:
Remove ignored files

pcnfernando committed:
Add eof

pcnfernando committed:
Add boundary validation sample

pcnfernando committed:
Add dop related samples

pcnfernando committed:
Add missing eof

pcnfernando committed:
Re-structure java samples

pcnfernando committed:
Update dop folder structure and include readme

pcnfernando committed:
Make prerequisites a heading

sameerajayasoma commented on declarative-data-processing-for-dop/main.bal:
```diff
@@ -0,0 +1,24 @@
+import ballerina/io;
+import ballerina/http;
```
This sample is same as https://github.com/ballerina-guides/integration-samples/blob/main/working_with_data/working_with_data.bal

Can we simply use the one we already have?

pcnfernando replied:
It's a little bit different. 
The following has group-by clause and a client invocation. 
If we are to update and use the same, we'll have to remove the client invocation. 
[working_with_data.bal](https://github.com/ballerina-guides/integration-samples/blob/main/working_with_data/working_with_data.bal) 

Or else, use exactly the same thing in https://github.com/ballerina-guides/integration-samples/blob/main/working_with_data/working_with_data.bal. wdyt?

pcnfernando replied:
Will be reusing the existing sample. Did a small amendment to the existing sample adding a group-by clause in https://github.com/ballerina-guides/integration-samples/pull/28/commits/b4b6d45633bc0d0cf74322a6617d846a663ecd50

Hope that's fine
```ballerina
import ballerina/http;
import ballerina/io;

type Country record {
    string country;
    int population;
    string continent;
    int cases;
    int deaths;
};

// Prints the top 10 countries having the highest case-fatality ratio grouped by continent.
public function main() returns error? {
    http:Client diseaseEp = check new ("https://disease.sh/v3");
    Country[] countries = check diseaseEp->/covid\-19/countries;

    json summary =
        from var {country, continent, population, cases, deaths} in countries
            where population >= 100000 && deaths >= 100
            let decimal caseFatalityRatio = (<decimal>deaths / <decimal>cases * 100).round(4)
            let json countryInfo = {country, population, caseFatalityRatio}
            order by caseFatalityRatio descending
            limit 10
            group by continent
            order by avg(caseFatalityRatio)
            select {continent, countries: [countryInfo]};
    io:println(summary);
}

```

sameerajayasoma commented on boundary-validation-for-dop/Readme.md:
```diff
@@ -0,0 +1,16 @@
+# Service Description:
+
+The service supports creating user records.
```
I would rename this directory as "data-validation"

pcnfernando replied:
Done with https://github.com/ballerina-guides/integration-samples/pull/28/commits/95e981d60f608cc016e76bfcc86fd5769c091174

sameerajayasoma commented:
Let's move most of these samples into a directory called "data-oriented-programming". We can remove the `-dop` with that.

pcnfernando committed:
Move dop samples under a new folder

pcnfernando committed:
Reuse exisiting work-with-data sample

gimantha commented on data-oriented-programming/declarative-data-processing/main.bal:
```diff
@@ -0,0 +1,24 @@
+
+public function main() returns error? {
+    // Perform data transformation using Ballerina's query language
+    json summary = from var {country, continent, population, cases, deaths} in <Country[]>check covidClient->/countries
```
@sameerajayasoma @pcnfernando  Why don't we promote having records instead of json type here?

pcnfernando replied:
The original thought was to do a transformation. 
Is your concern that even without explicit json conversion, it will be automatically handled at APIs?

gimantha replied:
Yeah.. We usually use the records inside the code to do data manipulation and use the json only via wire.

pcnfernando replied:
As I see it, there can be other scenarios even without the wire; for instance, storing in json files. Let's get insight from @sameerajayasoma as well

pcnfernando committed:
Improve query sample adding groupby

pcnfernando committed:
Resolve build failure with dist ver update

pcnfernando committed:
Remove .devcontainer and .gitingore files

pcnfernando committed:
Remove .gitignore files

pcnfernando committed:
Update PR build to run on nightly

pcnfernando commented:
Build failure is due to `group-by` feature not being available on U6 which the PR build runs with

sameerajayasoma commented on working_with_data/working_with_data.bal:
```diff
@@ -9,17 +9,20 @@ type Country record {
 };
 
-// Prints the top 10 countries having the highest case-fatality ratio.
+// Prints the top 10 countries having the highest case-fatality ratio grouped by continent.
```
Let's not change this example, at least right now. It is linked in Ballerina in action page as well.

pcnfernando replied:
Reverted and added another sample for the moment since old sample doesn't have group-by clause in a09d2a31e528e08ec9d858e59b89c4f0f4f7ae4e

sameerajayasoma commented on data-oriented-programming/xml-support/ballerina/main.bal:
```diff
@@ -0,0 +1,33 @@
+    string name = (person/<name>).data();
+    string age = (person/<age>).data();
+    string city = (person/**/<city>).data();
+    string country = (person/<address>/<country>).data();
```
Should we show that XML-to-record conversion capability as well? Similar to working with JSON example

pcnfernando replied:
The original example had it though. We discussed to remove record transformation which used `ballerina/xmldata;` import. Agree with showcasing that support is important.

Shall I revert it back to 

```ballerina
import ballerina/io;
import ballerina/xmldata;

// Define a SOAP payload
xml soapPayload = xml `<soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/">
    <soapenv:Body>
        <person>
            <name>John Doe</name>
            <age>30</age>
            <address>
                <city>New York</city>
                <country>USA</country>
            </address>
        </person>
    </soapenv:Body>
</soapenv:Envelope>`;

xmlns "http://schemas.xmlsoap.org/soap/envelope/" as ns;

type Address record {|
    string city;
    string country;
|};

public function main() returns error? {
    // Extract the SOAP payload
    xml xmlPayload = soapPayload/**/<ns:Body>;
    io:println(xmlPayload);

    // Navigate to the subcontext and extract the data
    xml person =  xmlPayload/<person>;

    string name = (person/<name>).data();
    string age = (person/<age>).data();
    string city = (person/**/<city>).data();

    // Extract the sub-xml and convert it to a record
    Address address = check xmldata:fromXml(person/<Address>);
    string country = address.country;

    io:println(string `Name: ${name}, Age: ${age}, City: ${city}, Country: ${country}`);
}

```

pcnfernando replied:
Updated in a09d2a31e528e08ec9d858e59b89c4f0f4f7ae4e

sameerajayasoma commented on data-oriented-programming/model-tabular-data/main.bal:
```diff
@@ -0,0 +1,40 @@
+// Define a type for tabular data
+type Employee record {|
+    readonly int id;
+    string name;
```
I would demonstrate a sample with just a single/simple key. That is a bit easier to understand for a new comer. WDYT?

pcnfernando replied:
Hmmm, what we had initially was that. But since java hashtable compound key support is not direct(will have to update the hash method to handle the compound fields) thought of having a compound key sample.
Agree with single key is much easier. wdyt?

sameerajayasoma replied:
It's okay. Let's keep this as it is for now.

sameerajayasoma commented on data-oriented-programming/model-optionality/ballerina/main.bal:
```diff
@@ -0,0 +1,31 @@
+type Person record {|
+    int id;
+    string name;
+    //optional typed field
```
I follow the following format when writing single-line code comments. 

`// Optional field`
- There exists a space right after `//` characters.
- Start the comment with an uppercase character.

I would like to see consistent format in code comments.

pcnfernando replied:
Updated in a09d2a31e528e08ec9d858e59b89c4f0f4f7ae4e

sameerajayasoma commented on data-oriented-programming/model-data-streams/main.bal:
```diff
@@ -0,0 +1,28 @@
+    float humidity;
+|};
+
+public function main() returns error? {
```
Take file_path as a command-line argument. We've used that pattern in other examples..

```
public function main(string filePath = "sensor_data.csv") returns error? {
........
}
```

pcnfernando replied:
Updated in a09d2a31e528e08ec9d858e59b89c4f0f4f7ae4e

pcnfernando committed:
Address review suggestions
- add new sample for query expressions
- update xml sample with record transformation
- update inline comments

pcnfernando committed:
Revert workflow change
- U7 is released

sameerajayasoma commented:
By the way, I didn't review the Java code in detail.

We need to mention the following in the README.md of each sample. 

We would highly appreciate community contribution to improving this Java code to make it the best for this scenario. Some thing like that.

pcnfernando committed:
Request contribution for java code

pcnfernando committed:
Update java xml to support record mapping

pcnfernando committed:
Update samples to honour 120 ll

pcnfernando committed:
Add missing eof
